dolphinscheduler-alert/src/main/java/org/apache/dolphinscheduler/alert/plugin/DolphinPluginClassLoader.java

```diff
@@ -0,0 +1,136 @@
+            ClassLoader spiClassLoader,
+            Iterable<String> spiPackages,
+            Iterable<String> spiResources) {
+        // plugins should not have access to the system (application) class loader
+        super(urls.toArray(new URL[urls.size()]), PLATFORM_CLASS_LOADER);
+        this.spiClassLoader = requireNonNull(spiClassLoader, "spiClassLoader is null");
+        this.spiPackages = ImmutableList.copyOf(spiPackages);
+        this.spiResources = ImmutableList.copyOf(spiResources);
+    }
+
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve)
+            throws ClassNotFoundException {
+        // grab the magic lock
+        synchronized (getClassLoadingLock(name)) {
+            // Check if class is in the loaded classes cache
+            Class<?> cachedClass = findLoadedClass(name);
+            if (cachedClass != null) {
+                return resolveClass(cachedClass, resolve);
+            }
+
+            // If this is an SPI class, only check SPI class loader
+            if (isSpiClass(name)) {
+                return resolveClass(spiClassLoader.loadClass(name), resolve);
+            }
+
+            // Look for class locally
+            return super.loadClass(name, resolve);
+        }
+    }
+
+    private Class<?> resolveClass(Class<?> clazz, boolean resolve) {
+        if (resolve) {
+            resolveClass(clazz);
+        }
+        return clazz;
+    }
+
+    @Override
+    public URL getResource(String name) {
+        // If this is an SPI resource, only check SPI class loader
+        if (isSpiResource(name)) {
+            return spiClassLoader.getResource(name);
+        }
+
+        // Look for resource locally
+        return super.getResource(name);
+    }
+
+    @Override
+    public Enumeration<URL> getResources(String name)
+            throws IOException {
+        // If this is an SPI resource, use SPI resources
+        if (isSpiClass(name)) {
+            return spiClassLoader.getResources(name);
+        }
+
+        // Use local resources
+        return super.getResources(name);
+    }
+
+    private boolean isSpiClass(String name) {
+        return spiPackages.stream().anyMatch(name::startsWith);
+    }
+
+    private boolean isSpiResource(String name) {
+        return spiResources.stream().anyMatch(name::startsWith);
+    }
+
+    private static String classNameToResource(String className) {
+        return className.replace('.', '/');
+    }
+
+    @SuppressWarnings("JavaReflectionMemberAccess")
+    private static ClassLoader findPlatformClassLoader() {
+        try {
+            // use platform class loader on Java 9
+            Method method = ClassLoader.class.getMethod("getPlatformClassLoader");
+            return (ClassLoader) method.invoke(null);
+        } catch (NoSuchMethodException ignored) {
+            // use null class loader on Java 8
+            return null;
+        } catch (IllegalAccessException | InvocationTargetException e) {
```

Comment:
Do we have a better way to determine the current Java version? OSUtils may have a similar method.

Reply:
I can not found the method to get the current Java version in OSUtils class， You can add it in your pr.
Thanks